File: JavaScript/프로그래머스/2/42626. 더 맵게/더 맵게.js
class MinHeap {
    constructor() {
        this.heap = [];
    }

    push(value) {
        this.heap.push(value);
        this.bubbleUp();
    }

    pop() {
        if (this.heap.length === 0) return null;
        const min = this.heap[0];
        const last = this.heap.pop();
        if (this.heap.length > 0) {
            this.heap[0] = last;
            this.bubbleDown();
        }
        return min;
    }

    bubbleUp() {
        let index = this.heap.length - 1;
        const current = this.heap[index];
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            const parent = this.heap[parentIndex];
            if (current >= parent) break;
            this.heap[parentIndex] = current;
            this.heap[index] = parent;
            index = parentIndex;
        }
    }

    bubbleDown() {
        let index = 0;
        const length = this.heap.length;
        const current = this.heap[0];

        while (true) {
            let leftIdx = 2 * index + 1;
            let rightIdx = 2 * index + 2;
            let swap = null;

            if (leftIdx < length && this.heap[leftIdx] < current) {
                swap = leftIdx;
            }

            if (
                rightIdx < length &&
                this.heap[rightIdx] < (swap === null ? current : this.heap[leftIdx])
            ) {
                swap = rightIdx;
            }

            if (swap === null) break;

            this.heap[index] = this.heap[swap];
            this.heap[swap] = current;
            index = swap;
        }
    }

    size() {
        return this.heap.length;
    }

    peek() {
        return this.heap[0];
    }
}

function solution(scoville, K) {
    const heap = new MinHeap();
    for (let s of scoville) heap.push(s);

    let count = 0;
    while (heap.size() >= 2 && heap.peek() < K) {
        const least = heap.pop();
        const secondLeast = heap.pop();
        const newScoville = least + secondLeast * 2;
        heap.push(newScoville);
        count++;
    }

    return heap.peek() >= K ? count : -1;
}

File: Python3/프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, K):
    answer = 0
    #  scoville을 힙 구조로 만들고
    #  heapq.heappop()을 사용해서 첫 번째 작은 값,
    #  또 heappop을 사용해서 두 번째 작은 값을 pop하고
    #  first + (second*2)해서 그 값을 heap에 push해주고 answer += 1 하기
    #  가장 작은 값(heap[0])이 K 이상이 될 때까지 반복
    #  더 이상 섞을 수 없으면 -1 return
    
    heapq.heapify(scoville)
    
    while scoville[0] < K:
        
        if len(scoville) < 2:
            return -1
        
        first = heapq.heappop(scoville)
        second = heapq.heappop(scoville)
        
        heapq.heappush(scoville, first + (second * 2))
        answer += 1

    return answer
